### ISSUE TITLE
Internal communication with TLS

### ISSUE THREAD

**Issue by jimmykarily** [epinio/epinio#130]
In addition to #62 and #63 we should also make sure all components talk to each other over secure channels with proper certificates. There are at least 2 ways to achieve this:

- Create certificates "manually" (e.g. with quarks secret) and setup https only services for components
- Use a service mesh (e.g. Linkerd or Istio) to automate this

Introducing a service mesh just for this may be an overkill. Needs some research before we decide. There may be other options which I didn't think of.

**Comment by agracey** [comment]
I'd suggest that this might not actually be needed as long as we build in a way that we could sit on top of Istio or Linkerd without issue.

**Comment by jimmykarily** [comment]
I tried Istio and seems to do the trick for internal mTLS between services. The steps to reproduce:

- Remove Traefik if it's running on the cluster (everything, deployments, services etc). There must be a way to have both Traefik and Istio running together but this may complicate things.
- Change gitea ingress and tekton ingress to have `kubernetes.io/ingress.class: istio` instead of `kubernetes.io/ingress.class: traefik` (since we no longer run traefik).
- Label the various namespaces to have the label `istio-injection=enabled` (in order to get sidecars next to every pod that acts as a backend for a service)
- Lock down to only mTLS on the entire mesh: https://istio.io/latest/docs/tasks/security/authentication/mtls-migration/#lock-down-mutual-tls-for-the-entire-mesh
- Change Tekton path to [what Istio understands](https://istio.io/latest/docs/tasks/traffic-management/ingress/kubernetes-ingress/#specifying-path-type): https://github.com/SUSE/carrier/blob/main/deployments/tekton.go#L363 :  `/` -> `/*`

that's it. To verify that no traffic is served over http, remove the `istio-injection=enabled` if you had that enabled, restart the api-ingress pod and get a shell to the new one (it shouldn't have a sidecar now). Try to `zypper in -y curl && curl -k http://gitea-http.gitea.svc.cluster.local:10080` . You should see: ```curl: (56) Recv failure: Connection reset by peer
```. Now add the `istio-injection=enabled` label back to the namespace, restart the api-ingress pod and get a shell to the api-ingress container (now it should have a sidecar next to it). Try the same curl command (after you install curl again). It should now succeed.

Having instio-injection enabled on `gitea` , `tekton-pipelines` and `carrier-workloads` namespaces, I got a successful `carrier push` of the sample app.

**Comment by jimmykarily** [comment]
For some reason the the application ingress is not accessible. It redirects to https and that one doesn't work. Gitea and Tekton are both accessible over `http` and I don't see any difference in the application ingress/service that forces it to redirect.

**Comment by jimmykarily** [comment]
Looking at the application container logs, the 301 seems to originate there, it's the application server itself that redirects. Maybe because of this? :

- https://github.com/paketo-buildpacks/php-web/pull/10
- https://github.com/paketo-buildpacks/php-web/pull/10

**Comment by jimmykarily** [comment]
Yes that's it. Can be disabled like this: https://github.com/SUSE/carrier/commit/feb1748f841230b6c9d8c908f34c026d8a0dc5b4

**Comment by jimmykarily** [comment]
Istio results:

- Makes mTLS relatively easy to implement between the various services.
- Doesn't seem to work when enabled on the gitea namespace. The gitea server init container fails to talk to the postgresql pod. The postgresql pod sidecar refuses the request with errors like: 
```
│ [2021-04-02T10:42:17.109Z] "- - -" 0 NR filter_chain_not_found - "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.42.0.173:5432 10.42.0.175:50262 - -                                                                                                │
```
  I spent quite some time trying to find a solution but to no avail. 

- Webhooks can't be automatically secured with mTLS because it's the kube api that talks to them (e.g. the app-ingress webhook) and the kube api doesn't go through istio, thus it doesn't have an appropriate certificate to do mTLS. Securing the kube webhooks needs manual setup of TLS.
- mTLS on the registry won't work either, for the same reason. Kube has to pull the images from the registry and has no cert for mTLS.

Overall, regarding encrypted communication between services, Istio makes some things easier but there are other things that either need serious Istio kung fu or manual handling.

**Comment by jimmykarily** [comment]
For the eirinix webhook server there seems to be a way to automatically setup a certificate using [the `SetupCertificate` option](https://github.com/cloudfoundry-incubator/eirinix/blob/master/manager.go#L392-L395)

**Comment by jimmykarily** [comment]
Useful FAQ: https://istio.io/latest/faq/security/#auth-mix-and-match

**Comment by jimmykarily** [comment]
Regarding Postgresql  issue:

[the reply to this question](https://stackoverflow.com/q/60679365) links to [this issue](https://github.com/istio/istio/issues/14743) -> which links to [this issue](https://github.com/istio/istio/issues/29761#issuecomment-758809821):

```
This was discussed on chat with @howardjohn, @mandarjog and others . Summary of our discussion, postgres uses starttls and this is causing issues with TLS origination. Istio would need a postgres filter for TLS origination to work with postgres.

In addition, there are no known benefits of doing TLS origination for postgres today. There is no improved security or improved visibility. As a result, it's recommended to simply start TLS from the application container directly. This way Istio at least sees the SNI.
```

**Comment by jimmykarily** [comment]
Looking at the default configuration of the bitnami postgresql helm chart (which is what the gitea chart uses), postgres tls is disabled.

Our issue may be related to this one:

- https://github.com/istio/istio/issues/27042 (solved here: https://github.com/istio/istio/pull/27105)

postgresql starts 2 pods: `gitea-postgresql` and `gitea-postgresql-headless` with the second one being a [headless service](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services). What doesn't fit is that this seems to be fixed on 1.7 and I'm using Istio 1.9.2 :thinking:

**Comment by agracey** [comment]
From my limited experience, Istio will cause us to not run in >4Gb... 

I'd also be interested in how it works for clusters that are using other service meshes.

I'll give a -1 to adding Istio as a requirement unless there's a very good reason to.

**Comment by agracey** [comment]
err, <4Gb

**Comment by jimmykarily** [comment]
These are valid concerns. I wanted to give Istio a try before we reject it. There is one thing Istio offers that is hard to beat. Automatic mTLS communication with the services not having to know anything about it. Of course it would make it more useful if it really worked in all cases (see my "Istio results" comment above).

The problem with mTLS is that all servers but also (even harder) all clients should be mTLS aware. Clients can't just take a URL and talk to it. This may be a problem for us because we don't have full control on all the clients and servers we use. For example gitea is talking to Tekton to trigger the staging pipeline. If gitea helm chart doesn't provide a way to setup mTLS communication for that, then we are out of luck.

Another problem is that even if we manually setup all communication to be mTLS/TLS, there is still a chance that we add a component/webhook/something that we forget to secure. With Istio you can set `PeerAuthentication` to be in `mode: STRICT` and be sure that you can't have non encrypted communication happening.

On the other hand, `mode STRICT` sounds more like a dream state which nobody seems to achieve but they are "getting there". Or maybe it's just me.

Next step is to give the manual setup a try and see where it's blocked (if at all). This story feels more like a spike than an actual implementation btw.

**Comment by jimmykarily** [comment]
I took Gitea as the first candidate for manual mTLS setup. Looking at the code [here](https://github.com/go-gitea/gitea/blob/7d7007dca75ab3a3a48b9f3fd7cc4350cc626870/services/webhook/deliver.go#L271), there is nothing that allows me to pass client certificates to the client. One would need to do something like this to get mTLS working: https://venilnoronha.io/a-step-by-step-guide-to-mtls-in-go#final-step---enable-mtls

We can still go with plain TLS though. Gitea even allows to set `InsecureSkipVerify` which means we can have self signed certificates if we need to.

If we had mTLS, no one other than Gitea would be able to trigger the staging pipeline. Without mTLS, I think (and I'm currently verifying) that it's possible for anyone in the cluster to make a request to the el-staging pod and trigger the staging pipeline for any of the applications.

**Comment by jimmykarily** [comment]
I pushed 2 instances of the sample app: `sample` and `sample-2`. I got a shell on the sample2 container and did a curl to the tekton webhook listener:

```
curl -v -X POST -H "Content-Type: application/json" --data '{ "head_commit": { "id": "fa81a503ff10ce3beeb5197c5262e6e6cc8b9b09" }, "repository": { "name": "sample", "full_name": "workspace/sample" }}' http://el-staging-li
stener.carrier-workloads:8080
```
This successfully triggered a staging job for the `sample` application. I didn't even have to use the `secret` defined here: https://github.com/SUSE/carrier/blob/main/internal/cli/clients/client.go#L1134

This is definitely a security concern and has to be fixed. mTLS would solve this by not allowing anyone else to talk to the webhook url. The use of some secret could solve this too (thus making plain TLS an option) but it doesn't seem to work right now. Either we are doing something wrong or authentication is not supported through the [tekton pipelines we are using](https://github.com/tektoncd/catalog/tree/main/task/buildpacks/0.2). I'm checking which is the case.

**Comment by jimmykarily** [comment]
I'm reading [the tekton docs](https://github.com/tektoncd/triggers/blob/main/docs/eventlisteners.md#multi-tenant-concerns) but there doesn't seem to be away to setup authentication (password or other secret) for the event listener. They only propose ways to limit what the listener can do on the cluster, not who can talk to the listener.

**Comment by agracey** [comment]
For auth through the ingress you can use Treafik's ForwardAuth functionality. This and CNI rules should let you lock it down a bit. I'm not sure how to do it internal to the cluster though.

**Comment by jimmykarily** [comment]
I created [this PR](https://github.com/SUSE/carrier/pull/251) to at least make use of the password we already have in place. I tried with the same steps as before and it's not possible to trigger a staging task without a password.

**Comment by wmorgan** [comment]
Maybe I'm late to the party, but if you want mTLS and are worried about the 4gb memory footprint, I would encourage you to try Linkerd. It will be significantly smaller at runtime than Istio, and mTLS is on by default between all meshed pods.

**Comment by agracey** [comment]
Are there any compatibility issues with including Linkerd if Istio already exists in the cluster?

**Comment by wmorgan** [comment]
An individual pod cannot have both sidecars, but otherwise there should be no compatibility issues. If you go down this path I'd love to hear how it goes... feedback welcome!

**Comment by jimmykarily** [comment]
I read:

```
As probes are initiated from the kubelet, which is not in the mesh, there is no identity and these requests are not mTLS’d, as denoted by the tls=no_tls_from_remote message.
```

(from here: https://linkerd.io/2.10/tasks/securing-your-service/)

So if one of the endpoints is not meshed, then mTLS is not applied. That works nicely for us because it means things can stay out of the mesh if we want to and that won't break anything. That is probably the explanation on why tekton could push the image to the registry.

My current plan:

- Verify the above assumption thaat mTLS is only applied when both sides are meshed.
- FInd a way to add TLS encryption to the registry (no mTLS, plain one-way TLS)
- Check in the linkerd viz dashboard what other namespace don't have automatic sidecar injection and decide if they need it or not (first quick look says they don't)
- [TBD] Add a flag in `epinio install` to setup linkerd. We can either install linkerd on the cluster ourselves, or simply add the needed annotations and tell the user that, if linkerd is installed, Epinio will play nice with it. Another option is to always install linkerd (even in development clusters) when it is not already installed (like we do for traefik). In that latter case, we probably just need a flag to skip the linkderd installation (the equivalent of this story for traefik: https://github.com/epinio/epinio/issues/380).

**Comment by jimmykarily** [comment]
```
Linkerd does not currently enforce mTLS. Any unencrypted requests inside the mesh will be opportunistically upgraded to mTLS. Any requests originating from inside or outside the mesh will not be automatically mTLS’d by Linkerd. This will be addressed in a future Linkerd release, likely as an opt-in behavior as it may break some existing applications.
```

https://linkerd.io/2.10/features/automatic-mtls/#caveats-and-future-work

I'm a bit confused on how the future release will `automatically mTLS` requests from outside the mesh :confused: . If it's outside the mesh, there is nothing you can do about it right? I'll enable automatic injection on the registry namespace too and try to see if there is a way to get this playing nicely with the kubelet not being in the mesh. For as long as linkerd doesn't enforce mTLS this should also work but we may be able to get it working even with mTLS enforced (in future linkerd releases).

**Comment by jimmykarily** [comment]
New issue:

```
│ service-catalog I0520 07:24:39.093791       1 hyperkube.go:192] Service Catalog version v0.3.1-dirty (built 2020-11-05T00:14:24Z)                                          │
│ service-catalog Error: failed to scale down controller manager: cannot get deployment service-catalog/service-catalog-catalog-controller-manager: Get https://10.43.0.1:44 │
│ 3/apis/apps/v1/namespaces/service-catalog/deployments/service-catalog-catalog-controller-manager: dial tcp 10.43.0.1:443: connect: connection refused                      │
│ linkerd-proxy time="2021-05-20T07:24:39Z" level=info msg="running version stable-2.10.2"                                                                                   │
│ linkerd-proxy [     0.006651s]  INFO ThreadId(01) linkerd2_proxy::rt: Using single-threaded proxy runtime                                                                  │
│ linkerd-proxy [     0.011413s]  INFO ThreadId(01) linkerd2_proxy: Admin interface on 0.0.0.0:4191                                                                          │
│ linkerd-proxy [     0.011436s]  INFO ThreadId(01) linkerd2_proxy: Inbound interface on 0.0.0.0:4143                                                                        ││ linkerd-proxy [     0.011440s]  INFO ThreadId(01) linkerd2_proxy: Outbound interface on 127.0.0.1:4140                                                                     ││ linkerd-proxy [     0.011443s]  INFO ThreadId(01) linkerd2_proxy: Tap interface on 0.0.0.0:4190                                                                            │
│ linkerd-proxy [     0.011446s]  INFO ThreadId(01) linkerd2_proxy: Local identity is clean-job-account.service-catalog.serviceaccount.identity.linkerd.cluster.local        │
│ linkerd-proxy [     0.011451s]  INFO ThreadId(01) linkerd2_proxy: Identity verified via linkerd-identity-headless.linkerd.svc.cluster.local:8080 (linkerd-identity.linkerd │
│ .serviceaccount.identity.linkerd.cluster.local)                                                                                                                            │
│ linkerd-proxy [     0.011454s]  INFO ThreadId(01) linkerd2_proxy: Destinations resolved via linkerd-dst-headless.linkerd.svc.cluster.local:8086 (linkerd-destination.linke │
│ rd.serviceaccount.identity.linkerd.cluster.local)                                                                                                                          │
│ linkerd-proxy [     0.029951s]  INFO ThreadId(02) daemon:identity: linkerd_app: Certified identity: clean-job-account.service-catalog.serviceaccount.identity.linkerd.clus │
│ ter.local                                                                                                                                                                  │
│ linkerd-init stream closed 
```

this happens when calling `epinio uninstall`. ` service-catalog-catalog-pre-delete-clean-job-dhkxw` which starts when uninstalling tekton (a helm hook?) tries to talk to the kube api and fails. That pod has a linkerd sidecar. Now, why didn't this work like talking to the registry did?

**Comment by jimmykarily** [comment]
Could this be the explanation? https://github.com/linkerd/linkerd2/issues/2118#issuecomment-457336357

**Comment by wmorgan** [comment]
This is a bit beyond my skillset to debug, but if you are stuck on this you could chat with the friendly folks on the [Linkerd Slack](https://slack.linkerd.io) or start a thread on the [Linkerd discussions forum](https://github.com/linkerd/linkerd2/discussions).

**Comment by rohitsakala** [comment]
*Progress* 

So, I tried out linkerd. I injected linkerd into tekton-staging, gitea, epinio, traefik, epinio-registry. I could confirm the following connections to work with mTLS.
1. Traefik -> Epinio
2. Epinio -> Gitea
3. Tekton -> Gitea
4. Tekton -> Epinio-registry

- Validating mTLS connections

-> We validated the mTLS connections using `linkerd tap` command. We validated `Tekton -> Epinio-registry` by checking the `tcp` requests on `wireshark` & `tshark` as `linkerd tap` was not detecting the requests from `Tekton -> Epinio-registry`. I think `linkerd tap` only detects HTTP/HTTP2 connections (Need to file an issue)

The connections which don't work as mTLS are

1. Epinio CLI -> Traefik
2. Kubelet -> Epinio-Registry 

Things to check are service-catalog and cert-manager namespaces. I think epinio components interact using CRD's with these components. I need to confirm this. If so, we can ignore injecting linkerd into these components.

Cons

The clone, stage, and run pods after completion be in the `CannotRunContainer` error state. This is a valid behavior as `Tekton` replaces all images of sidecars with `nop` image [0].


[0] https://github.com/tektoncd/pipeline/issues/727#issuecomment-494388473
